### PR TITLE
[Merged by Bors] - fix: correct `String.instLinearOrderString`

### PR DESCRIPTION
**Pull request by chabulhwi**
Use the optimized `compare` function and provide a manual `compare_eq_compareOfLessAndEq` field.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/String/Basic.lean [hunk 1]
```diff
@@ -226,6 +226,24 @@ theorem popn_empty {n : ℕ} : "".popn n = "" := by
...
+  · induction h with
+    | nil b bs => exact Lex.nil
+    | @head a as b bs hab => apply Lex.rel; assumption
+    | @tail a as b bs hab hba _ ih =>
+      have heq : a = b := _root_.le_antisymm (le_of_not_lt hba) (le_of_not_lt hab)
+      subst b; apply Lex.cons; assumption
+  · induction h with
+    | @nil a as => apply List.lt.nil
+    | @cons a as bs _ ih => apply List.lt.tail <;> simp [ih]
+    | @rel a as b bs h => apply List.lt.head; assumption
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/String/Basic.lean:243]: Note that there's nothing specific about `Char` in this proof; the following works:
```lean
theorem List.lt_eq_lt [LinearOrder α] (l₁ l₂ : List α) : List.lt l₁ l₂ ↔ l₁ < l₂ := by
  dsimp only [List.LT', List.lt]
  constructor
  · intro h
    induction h with
    | nil b bs => exact Lex.nil
    | @head a as b bs hab => apply Lex.rel; assumption
    | @tail a as b bs hab hba _ ih =>
      have heq : a = b := _root_.le_antisymm (le_of_not_lt hba) (le_of_not_lt hab)
      subst b; apply Lex.cons; assumption
  · intro h
    induction h with
    | @nil a as => apply List.lt.nil
    | @cons a as bs _ ih => apply List.lt.tail <;> simp [ih]
    | @rel a as b bs h => apply List.lt.head; assumption

theorem List.lt_iff_lex_char_lt (l l' : List Char) : List.lt l l' ↔ List.Lex Char.lt l l' :=
  List.lt_eq_lt l l'
```
(Why do we override the LT instance on List anyway?)

Comment (eric-wieser) [reply Mathlib/Data/String/Basic.lean:243]: Some of this might be a mess made by porting stuff that std4 already had a competing implementation of.

Comment (chabulhwi) [reply Mathlib/Data/String/Basic.lean]: > Note that there's nothing specific about `Char` in this proof; the following works:
I changed the theorem `List.lt_iff_lex_char_lt` to `List.lt_iff_lex_lt`.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**